runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,431 @@
+
+    /**
+     * Start the actual corfu transaction. Ensure there isn't one already in the same thread.
+     */
+    private void txBeginInternal() {
+        if (TransactionalContext.isInTransaction()) {
+            log.error("Cannot start new transaction in this thread without ending previous one");
+            throw new TransactionAlreadyStartedException(TransactionalContext.getRootContext().toString());
+        }
+        this.txnStartTime = System.nanoTime();
+        Transaction.TransactionBuilder transactionBuilder = this.objectsView
+                .TXBuild()
+                .type(TransactionType.WRITE_AFTER_WRITE);
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
+        transactionBuilder.build().begin();
+    }
+
+    /**
+     * Commit the transaction.
+     * For a transaction that only has write operations, this method will start and end
+     * a corfu transaction keeping the "critical" section small while batching up all updates.
+     * For a transaction that has reads, this will end the transaction.
+     * The commit returns successfully if the write transaction was committed.
+     * Otherwise this throws a TransactionAbortedException with a cause.
+     * The cause and the caller's intent of the transaction can determine if this aborted
+     * Transaction can be retried.
+     * @return - address at which the commit of this transaction occurred.
+     */
+    public long commit() {
+        operations.forEach(Runnable::run);
+        long commitAddress = Address.NON_ADDRESS;
+        try {
+            commitAddress = this.objectsView.TXEnd();
+        } catch (Exception ex) {
+            tablesInTxn.forEach(t -> t.getMetrics().numTxnAborts.incrementAndGet());
+            tablesInTxn.clear();
+            throw ex;
+        }
+        long timeElapsed = System.nanoTime() - txnStartTime;
+        switch (txnType) {
+            case READ_ONLY:
+                tablesInTxn.forEach(t -> t.getMetrics().readOnlyTxnTimes.set(timeElapsed));
+                break;
+            case WRITE_ONLY:
+                tablesInTxn.forEach(t -> t.getMetrics().writeOnlyTxnTimes.set(timeElapsed));
+                break;
+            case READ_WRITE:
+                tablesInTxn.forEach(t -> t.getMetrics().readWriteTxnTimes.set(timeElapsed));
+                break;
+            default:
+                log.error("UNKNOWN TxnType!!");
+                break;
+        }
+        tablesInTxn.clear();
+        operations.clear();
+        return commitAddress;
+    }
+
+    /**
+     * Explicitly abort a transaction in case of an external failure
+     */
+    public void txAbort() {
+        operations.clear();
+        tablesInTxn.clear();
+        if (TransactionalContext.isInTransaction()) {
+            this.objectsView.TXAbort();
+            tablesInTxn.forEach(t -> t.getMetrics().numTxnAborts.incrementAndGet());
+        }
```

Comment:
The order of these statements seems incorrect, tablesInTxn is  cleared on line 406, but then it's iterated on line 409

Reply:
done